tests/utils/tensorflow/test_models.py

```diff
@@ -0,0 +1,107 @@
+        ),
+        (
+            {"a": np.array([1, 2]), "b": {"c": np.array([3, 1])}},
+            {"a": np.array([5, 6]), "b": {"c": np.array([2, 4])}},
+            {"a": np.array([1, 2, 5, 6]), "b": {"c": np.array([3, 1, 2, 4])}},
+        ),
+    ],
+)
+def test_merging_batch_outputs(
+    existing_outputs: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+    new_batch_outputs: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+    expected_output: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+):
+
+    predicted_output = _merge_batch_outputs(existing_outputs, new_batch_outputs)
+
+    def test_equal_dicts(
+        dict1: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+        dict2: Dict[Text, Union[np.ndarray, Dict[Text, np.ndarray]]],
+    ):
+        assert dict2.keys() == dict1.keys()
+        for key in dict1:
+            val_1 = dict1[key]
+            val_2 = dict2[key]
+            assert type(val_1) == type(val_2)
+
+            if isinstance(val_2, np.ndarray):
+                assert np.array_equal(val_1, val_2)
+
+            elif isinstance(val_2, dict):
+                test_equal_dicts(val_1, val_2)
+
+    test_equal_dicts(predicted_output, expected_output)
+
+
+@pytest.mark.parametrize(
+    "batch_size, number_of_data_points, expected_number_of_batch_iterations",
+    [(2, 3, 2), (1, 3, 3), (5, 3, 1),],
+)
+def test_batch_inference(
+    batch_size: int,
+    number_of_data_points: int,
+    expected_number_of_batch_iterations: int,
+):
+    model = RasaModel()
+
+    def batch_predict(batch_in: np.ndarray):
+
+        dummy_output = batch_in[0]
+        output = {
+            "dummy_output": dummy_output,
+            DIAGNOSTIC_DATA: tf.constant(np.array([[1, 2]]), dtype=tf.int32),
+        }
+        return output
+
+    # Monkeypatch batch predict so that run_inference interface can be tested
+    model.batch_predict = batch_predict
+
+    # Create dummy model data to pass to model
+    model_data = RasaModelData(
+        label_key="label",
+        label_sub_key="ids",
+        data={
+            "text": {
+                "sentence": [
+                    FeatureArray(
+                        np.random.rand(number_of_data_points, 2),
+                        number_of_dimensions=2,
+                    ),
+                ]
+            }
+        },
+    )
+    output = model.run_inference(model_data, batch_size=batch_size)
+
+    # Firstly, the number of data points in dummy_output should be equal
+    # to the number of data points sent as input.
+    assert output["dummy_output"].shape[0] == number_of_data_points
+
+    # Secondly, the number of data points inside diagnostic_data should be
+    # equal to the number of batches passed to the model because for every
+    # batch passed as input, it would have created a
+    # corresponding diagnostic data entry.
+    assert output[DIAGNOSTIC_DATA].shape == (expected_number_of_batch_iterations, 2)
```

Comment:
Maybe this is nitpicking, but in the real model code `output[DIAGNOSTIC_DATA]` is always a dict (usually containing numerical arrays somewhere down the line), not a numerical object itself. Perhaps there's value in changing the diagnostic data in the monkey-patched `_batch_predict` to be closer to reality in this sense. What do you think?

Reply:
I think that's a bit of an overkill because `test_merging_batch_outputs` already tests for the nested dict structure you mentioned. I actually wanted to not use `DIAGNOSTIC_DATA` as one of the keys of the output of `batch_predict` but then `rasa_predict` does something specific for `DIAGNOSTIC_DATA` and hence had to include it. Lmk what you think.

Reply:
Ah, I see! You made me realise that my PR actually [added code in `rasa_predict` which now effectively requires `DIAGNOSTIC_DATA` to be present](https://github.com/RasaHQ/rasa/commit/ca66e3494524d2ce3a30c3b815da27c4d0a558d8#diff-1b0d53e90a7f1dcca1113cda22c58c98685004f8dfc7ba3d4e5bf9cee78e6a19R255). That's a mistake I guess, and should be fixed. Then you won't have to use `DIAGNOSTIC_DATA` in your test 🙂  Sounds good?

Reply:
Sounds good. Do you have a fix for it in mind already?

Reply:
Yeah, the lines I added over there have to be conditioned on `DIAGNOSTIC_DATA` being present in the outputs dict. Simple 🙂  If you think it can go in your PR, feel free to make a fix; otherwise, I'll create a mini-PR.

Reply:
Done :+1

Reply:
Awesome, thanks!